### PR TITLE
ROSAENG-61174 | test: complete remaining pkg/aws coverage

### PR DESCRIPTION
**Pull request by olucasfreitas**
## PR Summary
Add focused unit tests for previously untested ROSA-specific business logic in `pkg/aws`, covering IAM role and policy lifecycle, IAM discovery and upgrade-selection logic, STS validation, CloudFormation stack operations, S3 bucket lifecycle, Secrets Manager helpers, EC2/VPC lookup helpers, service quota validation, and SCP permission simulation.

JIRA: [ROSAENG-61174](https://redhat.atlassian.net/browse/ROSAENG-61174)

## Detailed Description of the Issue
Several ROSA-specific `pkg/aws` behaviors were either untested or only indirectly covered, which made regressions harder to catch in areas such as IAM role creation/update/deletion, policy compatibility checking, role discovery and upgrade-selection, CloudFormation stack lifecycle handling, S3 bucket management, Secrets Manager operations, EC2/VPC subnet and availability zone lookups, service quota validation, and SCP permission simulation. This PR adds targeted automated coverage for those paths so future changes in AWS-facing logic can be validated more reliably.

The SCP simulation path (`checkPermissionsUsingQueryClient`) previously constructed a real IAM client via `iam.NewFromConfig()`, bypassing the injected mockable interface. A small production change adds `SimulatePrincipalPolicy` to the `IamApiClient` interface and uses the injected client instead, making the simulation path testable without widening the production API.

## Related Issues and PRs
- Jira: [ROSAENG-61174](https://redhat.atlassian.net/browse/ROSAENG-61174)
- Fixes: `#`
- Related PR(s): #3299
- Related design/docs: N/A

## Type of Change
- [ ] feat - adds a new user-facing capability.
- [ ] fix - resolves an incorrect behavior or bug.
- [ ] docs - updates documentation only.
- [ ] style - formatting or naming changes with no logic impact.
- [ ] refactor - code restructuring with no behavior change.
- [x] test - adds or updates tests only.
- [ ] chore - maintenance work (tooling, housekeeping, non-product code).
- [ ] build - changes build system, packaging, or dependencies for build output.
- [ ] ci - changes CI pipelines, jobs, or automation workflows.
- [ ] perf - improves performance without changing intended behavior.

## Previous Behavior
Key ROSA-specific AWS workflows and helper paths had limited or no direct automated coverage, including:
- IAM role lifecycle (`EnsureRole`, `EnsurePolicy`, `ForceEnsurePolicy`, `ValidateRoleNameAvailable`, `PutRolePolicy`, `AttachRolePolicy`, `DeleteInlineRolePolicies`)
- IAM discovery and upgrade-selection (`FindRoleARNs*`, `ValidateAccountRoleVersionCompatibility`, `FindPolicyARN`, `HasHostedCPPolicies`, `HasManagedPolicies`, `UpdateTag`, `AddRoleTag`, `GetAccountRoleVersion`, `IsAdminRole`)
- STS validation (`HasPermissionsBoundary`)
- Shared client helpers (`CheckRoleExists`, `GetRoleByARN`, `GetRoleByName`)
- CloudFormation wrappers (`GetCFStack`, `DescribeCFStackResources`)
- S3 lifecycle (`CreateS3Bucket`, `DeleteS3Bucket`)
- Secrets Manager (`CreateSecretInSecretsManager`, `DeleteSecretInSecretsManager`)
- EC2/VPC helpers (`DescribeAvailabilityZones`, `IsLocalAvailabilityZone`, `GetSubnetAvailabilityZone`, `GetVPCSubnets`)
- Service quotas (`GetIAMServiceQuota`)
- SCP simulation (`checkPermissionsUsingQueryClient`)

## Behavior After This Change
All listed functions now have focused automated tests covering expected behavior and relevant edge/error cases. The SCP simulation path is now testable via the injected IAM mock. There is no intended user-facing behavior change.

New test coverage summary:
- `policies_test.go`: ~52 new specs (IAM mutators + discovery/upgrade)
- `client_test.go`: ~30 new specs (role lookups, S3, Secrets, EC2/VPC)
- `policy_document_test.go`: 6 new specs (SCP simulation)
- `sts_test.go`: 3 new specs (HasPermissionsBoundary)
- `helpers_test.go`: 4 new specs (GetJumpAccount, ComputeOperatorRoleArn)
- `cloudformation_test.go`: 5 new specs (GetCFStack, DescribeCFStackResources)
- `quota_test.go`: 2 new specs (GetIAMServiceQuota)

### Preconditions
- Go toolchain installed (Go 1.26.3+ or GOTOOLCHAIN=auto)
- Repository cloned with hooks installed

### Test Steps
1. `GOTOOLCHAIN=auto go test ./pkg/aws/... -count=1 -cover`
2. `GOTOOLCHAIN=auto go vet ./pkg/aws/...`
3. `make test`
4. `make lint`

### Expected Results
- All `pkg/aws` tests pass with 40%+ coverage
- `go vet` clean
- `make test` passes (full repo)
- `make lint` passes (0 issues)

## Proof of the Fix
- Pre-push checks passed (format, build, lint, changed-files coverage, unit/integration tests)
- `make fmt`: clean
- `make lint`: 0 issues
- `make test`: all packages pass
- `go test ./pkg/aws/... -count=1 -cover`: 40.0% coverage, 392 specs pass

## Breaking Changes
- [x] No breaking changes
- [ ] Yes, this PR introduces a breaking change (describe impact and migration plan below)

### Breaking Change Details / Migration Plan
N/A

## Developer Verification Checklist
- [x] Commit subject/title follows `[JIRA-TICKET] | [TYPE]: <MESSAGE>`.
- [x] PR description clearly explains both **what** changed and **why**.
- [x] Relevant Jira/GitHub issues and related PRs are linked.
- [x] `make install-hooks` has been run in this clone.
- [x] Tests were added/updated where appropriate.
- [ ] I manually tested the change.
- [x] `make test` passes.
- [x] `make lint` passes.
- [ ] `make rosa` passes.
- [ ] Documentation or repo-local agent guidance was added/updated where appropriate.
- [x] Any risk, limitation, or follow-up work is documented.

[ROSAENG-61174]: https://redhat.atlassian.net/browse/ROSAENG-61174?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ROSAENG-61174]: https://redhat.atlassian.net/browse/ROSAENG-61174?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved IAM permission checks by using the already-configured IAM client when simulating policies, leading to more consistent results and clearer error propagation.
  * Enhanced handling of permission-evaluation edge cases, including implicitly denied actions and optional region context during simulation.

* **Tests**
  * Expanded coverage for S3 bucket and object deletion flows, Secrets Manager create/delete, CloudFormation stack queries, and broader IAM role/policy assurance workflows.
  * Added additional tests for IAM quotas, EC2 availability zone/VPC+subnet helpers, jump account/role ARN generation, and permissions-boundary detection.

### CODE REVIEW

**Comment by amandahla** [comment]
@coderabbitai Review this PR with a brief, technical, neutral tone. Prefer existing `pkg/aws` patterns; do not invent new conventions in a coverage-focused change. High-signal only — no style nits.

Check and report on:

1) **Branch coverage** — Are new specs exhaustive? Call out real gaps only (e.g. `GetJumpAccount` FedRAMP / `local-proxy` / `crc`; IAM discovery pagination/truncation/tag-version edges; S3/CF/Secrets multi-step early returns). Distinguish “nice to have” vs “should block”.

2) **Redundant / map-lock tests** — Does `GetJumpAccount` only restate `JumpAccounts` constants (weak)? Is `ComputeOperatorRoleArn` exercising real logic (path/GovCloud)? Are `IsPolicyCompatible` additions extending the existing describe appropriately?

3) **Overlap with existing tests** — What was already covered in `pkg/aws` vs net-new gaps (`FindRoleARNs*`, Ensure*, S3/Secrets/EC2, SCP simulation)? Flag only clear duplication.

4) **Imports** — Flag unnecessary aliases vs nearby files: `gomock "go.uber.org/mock/gomock"` (default name is already `gomock`; mixed with unaliased imports in `pkg/aws`). `awsSdk` for `aws-sdk-go-v2/aws` when `aws` is already imported (prefer one name). Match dominant package style; don’t prescribe a new alias convention.

5) **`policy_document.go` production change** — Confirm intent (inject `iamClient` so SCP simulation is mockable vs `iam.NewFromConfig`). Assess user impact / behavioral equivalence. Note this is a small production seam in a mostly-test PR; justified if equivalent, but should be called out explicitly. 

**Consistency bar (important):** Stay aligned with established `pkg/aws` test/import/mock patterns. Do not endorse new one-off styles that later coverage PRs would copy. Same concern as https://github.com/openshift/rosa/pull/3368#discussion_r3595743058 — coverage work should follow existing seams, not expand conventions.

Output: short verdict + bullets for actionable findings only. If something matches existing package practice, say so and skip.

**Comment by amandahla** [comment]
@coderabbitai this is my checklist, add comments if needed so @olucasfreitas can review it with more details. Add filename and line number.

Please check:

- [ ] Confirm the `policy_document.go` seam (`iam.NewFromConfig` → injected `iamClient`) has no user-facing behavior change for normal clients.
- [ ] Confirm SCP simulation tests actually pass against the injected mock (paginator call shape / expectations).
- [ ] Add coverage for `DeleteS3Bucket` when `DeleteObject` fails inside `emptyS3Bucket` (early return path).
- [ ] Revisit `GetJumpAccount` tests: they mostly restate `JumpAccounts` map literals. Prefer a real branch test for `fedramp.Enabled()`, or drop the map-lock specs.
- [ ] Confirm new coverage doesn’t duplicate existing describes (esp. around `IsPolicyCompatible`).
- [ ] Confirm no new import-alias convention (`aws` vs `awsSdk`, unnecessary `gomock` alias) is being expanded beyond existing `pkg/aws` patterns.
- [ ] Spot-check that failure-path tests assert meaningful wrapped errors where the production code adds context (not only `HaveOccurred()`).

File: pkg/aws/policy_document_test.go [hunk 1]
```diff
@@ -321,4 +330,159 @@ var _ = Describe("PolicyDocument", func() {
...
+		})
+
+		It("returns true when all actions are allowed", func() {
+			doc := &PolicyDocument{
+				Statement: []PolicyStatement{
+					{Effect: "Allow", Action: []interface{}{"s3:GetObject", "s3:PutObject"}},
+				},
+			}
+
+			mockIamAPI.EXPECT().SimulatePrincipalPolicy(gomock.Any(), gomock.Any(), gomock.Any()).
```
[the remark this answers is not in the document]
Comment (olucasfreitas) [reply pkg/aws/policy_document_test.go:361]: Checked this against the current paginator call shape. `NextPage()` still forwards an option function (`addIsPaginatorUserAgent`), so the gomock call is three-argument in practice. `GOTOOLCHAIN=auto go test ./pkg/aws/... -count=1` is passing with the current matcher shape, so I left this one unchanged.

File: pkg/aws/client_test.go [hunk 2]
```diff
@@ -953,4 +957,559 @@ var _ = Describe("Client", func() {
...
+			mockS3API.EXPECT().PutBucketTagging(gomock.Any(), gomock.Any(), gomock.Any()).
+				Return(nil, fmt.Errorf("tagging error"))
+
+			err := client.CreateS3Bucket(bucketName, DefaultRegion)
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).To(ContainSubstring("tagging error"))
+		})
+	})
+
+	Context("DeleteS3Bucket", func() {
```
[the remark this answers is not in the document]
Comment (olucasfreitas) [reply pkg/aws/client_test.go:1094]: Added the missing `DeleteObject` failure case in `pkg/aws/client_test.go`. The `DeleteS3Bucket` matrix now covers the early-return path where object deletion fails before `DeleteBucket` is reached.

File: pkg/aws/helpers_test.go [hunk 3]
```diff
@@ -1045,3 +1046,83 @@ var _ = Describe("resolveSTSRole", func() {
 		})
 	})
 })
+
+var _ = Describe("GetJumpAccount", func() {
+	When("env is production", func() {
+		It("Should return the production jump account", func() {
+			Expect(GetJumpAccount("production")).To(Equal("710019948333"))
```
[the remark this answers is not in the document]
Comment (olucasfreitas) [reply pkg/aws/helpers_test.go]: Reworked the `GetJumpAccount` tests in `pkg/aws/helpers_test.go` to cover both the standard path and the FedRAMP path via `fedramp.Enable()` / `fedramp.Disable()`, instead of only hardcoding default-map values.

**Comment by olucasfreitas** [comment]
Current review pass is complete.

Addressed in code:
- `pkg/aws/client_test.go`: added the `DeleteS3Bucket` early-return case where `DeleteObject` fails before `DeleteBucket`
- `pkg/aws/helpers_test.go`: replaced the `GetJumpAccount` map-lock assertions with branch coverage for both the standard and FedRAMP paths
- `pkg/aws/client_test.go`: added the `DeleteSecretInSecretsManager` fallthrough case where `DescribeSecret` fails with a non-`ResourceNotFoundException` error
- `pkg/aws/policies_test.go`: tightened the incompatible `EnsurePolicy` path to assert that non-default version `v2` is the one deleted

Verified and intentionally left unchanged:
- `pkg/aws/policy_document_test.go`: the current three-argument `SimulatePrincipalPolicy` expectations match the actual paginator call shape after rerunning `GOTOOLCHAIN=auto go test ./pkg/aws/... -count=1`

Latest branch head: `615c750`
- pre-push checks passed again: format, build, lint, changed-files coverage, and unit/integration tests

File: pkg/aws/policy_document_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (amandahla) [reply pkg/aws/policy_document_test.go:361]: @coderabbitai learn and resolve the comment
